### PR TITLE
Start/Stop plugin

### PR DESCRIPTION
**Pull request by dirksammel**
Adds plugin that sets demand to zero if no pending jobs are present on monitored partition. https://github.com/MatterMiners/cobald/issues/104

### CODE REVIEW

**Comment by maxfischer2781** [comment]
Thanks for this pull request! Since this is the first PR to this repo, it might take a bit until we have the entire review process going.

I've triggered all existing checks; they should rerun automatically when you push again.  
It looks like the test failures are due to some imports still assuming this is part of the `cobald` package itself – using absolute imports should fix this.

File: cobald_hep_plugins/stopper.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+from cobald.daemon import service
+
+
+@service(flavour=asyncio)
+class Stopper(PoolDecorator):
+    """
+    Decorator that sets demand to 0 if the partition has no pending jobs
+
+    :param target: the pool
+    :param script: script that checks for pending jobs on the partition
```
Comment (maxfischer2781) [inline cobald_hep_plugins/stopper.py:16]: As far as I can tell, the code expects to be given the path to the script (not the name, code, ...).
```suggestion
    :param script: path to script that checks for pending jobs
```

File: cobald_hep_plugins/stopper.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+import asyncio
+
+from cobald.daemon import service
+
+
+@service(flavour=asyncio)
+class Stopper(PoolDecorator):
+    """
+    Decorator that sets demand to 0 if the partition has no pending jobs
```
Comment (maxfischer2781) [inline cobald_hep_plugins/stopper.py:13]: I think you can drop restricting this just to a partition from here and later on – the code/plugin is much more generally useful.
```suggestion
    Decorator that sets demand to 0 if there are no pending jobs
```

File: cobald_hep_plugins/stopper.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    If there are pending jobs on the partition, the demand is not modified.
+    The demand is set to 0 as long as no pending jobs are detected.
+
+    The default interval is 300 (5 minutes). The script has to be specified.
+    """
+
+    @property
+    def demand(self) -> float:
+        return self._demand
```
Comment (maxfischer2781) [inline cobald_hep_plugins/stopper.py:27]: Since the demand is just passed along unchanged, there is no need to store it on the class. You can fetch it from the target.
```suggestion
    @property
    def demand(self) -> float:
        return self.target.demand
```

File: cobald_hep_plugins/stopper.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    def demand(self, value: float):
+        self._demand = self._condition_slurm(value)
+        self.target.demand = self._demand
+
+    def _condition_slurm(self, value):
+        """Return 0 if there are no pending jobs, otherwise pass `value`"""
+        if self.n_pend_jobs == 0:
+            return 0
+        else:
+            return value
```
Comment (maxfischer2781) [inline cobald_hep_plugins/stopper.py:39]: You can shorten this to directly evaluating the condition here – it's not used anywhere else, so re-use isn't a concern. Not sure if a ternary `if` or multiple lines is more readable.
```suggestion
        self.target.demand = value if self.n_pend_jobs else 0
```

File: cobald_hep_plugins/stopper.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    async def run(self):
+        """Retrieve the number of pending jobs"""
+        while True:
+            proc = await asyncio.create_subprocess_shell(
+                f". {self.script}",
+                stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.PIPE,
+            )
+            stdout, stderr = await proc.communicate()
+            self.n_pend_jobs = int(stdout.decode("ascii").strip())
```
Comment (maxfischer2781) [inline cobald_hep_plugins/stopper.py:50]: AFAIK the `.strip()` is not needed – `int` should already ignore whitespace.
```suggestion
            self.n_pend_jobs = int(stdout.decode("ascii"))
```

File: cobald_hep_plugins/stopper.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    def __init__(
+        self,
+        target: Pool,
+        script: str = "",
+        interval: int = 300,
+    ):
+        super().__init__(target)
+        enforce(interval > 0, ValueError("interval must be positive"))
+        enforce(script != "", ValueError("script must be specified"))
+        self._demand = target.demand
```
Comment (maxfischer2781) [inline cobald_hep_plugins/stopper.py:62]: If you directly use `self.target.demand` in the property, you can remove this attribute.

File: docs/plugins/stopper.rst [hunk 7]
```diff
@@ -0,0 +1,22 @@
+#################
+Start/Stop Plugin
+#################
+
+.. py:module:: cobald_hep_plugins.stopper
+    :synopsis: Stops the booting of new drones if there are no pending jobs on the monitored partition
```
Comment (maxfischer2781) [inline docs/plugins/stopper.rst:6]:
```suggestion
    :synopsis: Stops the booting of new drones if there are no pending jobs
```

File: tests/test_stopper.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            Stopper(pool)
+
+    def test_running(self):
+        pool = MockPool()
+        stopper = Stopper(pool, script="test.sh")
+
+        for pend_jobs in (2, 7, 150, 5000):
+            stopper.n_pend_jobs = pend_jobs
+            for value in (0, 1, 5, 10, 1000):
+                assert stopper._condition_slurm(value) == value
```
Comment (maxfischer2781) [inline tests/test_stopper.py:25]: If you apply the code changes suggested, you need to set the demand and then check it here. Something like this:
```suggestion
               stopper.demand = value
               assert stopper.demand == value
```

File: tests/test_stopper.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+    def test_idle(self):
+        pool = MockPool()
+        stopper = Stopper(pool, script="test.sh")
+
+        stopper.n_pend_jobs = 0
+
+        for value in (0, 1, 5, 10, 1000):
+            assert stopper._condition_slurm(value) == 0
```
Comment (maxfischer2781) [inline tests/test_stopper.py:34]: Like above, but asserting for equality with `0`.

**Review by maxfischer2781 — CHANGES_REQUESTED** [review]
Thanks a lot for fixing the import setup! I've added some code suggestions to shorten the code a bit, and especially make the description/wording more general – your plugin can be used for many more use cases!

**Comment by dirksammel** [comment]
Thanks for the suggested changes, I implemented them all.

**Comment by maxfischer2781** [comment]
The PyPy failures are due to dependencies failing to build, let's ignore them for this PR.

**Review by maxfischer2781 — APPROVED** [review]
Thanks, this looks good to go!

I'll merge changes and prepare the first release over the next few days.
